### PR TITLE
Support set and atomic mode

### PR DESCRIPTION
**Pull request by abdullah-arif-swe**
Creating to report coverage correct for atomic and set modes

### CODE REVIEW

File: go.mod [hunk 1]
```diff
@@ -14,6 +15,6 @@ require (
 )
 
 retract (
-    v0.3.0 // Version with problems.
```
Comment (abdullah-arif-swe) [inline go.mod:17]: auto diff - `go mod tidy` did this - I can revert if you prefer

File: gocovparser/core.go [hunk 2]
```diff
@@ -27,145 +27,73 @@ var parseLineRegex = regexp.MustCompile(
 	`(?P<host>[^\/]*)\/` + // github.com
 		`(?P<owner>[^\/]*)\/` + // heynemann
 		`(?:(?P<repo>[^\/]*)\/)?` + // gocovparser
-		`(?P<path>.*)\:` + // gocovparser/core.go
```
Comment (abdullah-arif-swe) [inline gocovparser/core.go:30]: This now works on file name instead - so the regex is simpler

File: gocovparser/core.go [hunk 3]
```diff
@@ -27,145 +27,73 @@ var parseLineRegex = regexp.MustCompile(
...
 
-	return ParseLines(lines)
-}
-
-// ParseLines of a coverage file contents from go tests.
-func ParseLines(coverageData []string) ([]Coverage, error) {
-	data := filterEmpty(coverageData)
-	if len(data) == 0 {
-		return []Coverage{}, nil
+	profiles, err := cover.ParseProfilesFromReader(strings.NewReader(coverageData))
```
Comment (abdullah-arif-swe) [inline gocovparser/core.go:30]: Simplified by relying on the go package, so the coverage is consistent with the Go tools

File: gocovparser/core.go [hunk 4]
```diff
@@ -27,145 +27,73 @@ var parseLineRegex = regexp.MustCompile(
...
-		return -1, errors.Wrapf(err, "parsing int: '%s'", val)
-	}
-
-	return int(result), nil
-}
-
 // GroupCoverage in the specified groups.
 func GroupCoverage(items []Coverage, groups ...ParseGroup) (ParseGroupResult, error) {
-	result := map[string]map[string]float64{}
+	result := make(map[string]map[string]float64)
```
Comment (abdullah-arif-swe) [inline gocovparser/core.go:67]: Following https://github.com/uber-go/guide/blob/master/style.md#initializing-maps
Can revert if you prefer old style

File: gocovparser/core_test.go [hunk 5]
```diff
@@ -41,11 +41,11 @@ func TestCanGroupCoverageData(t *testing.T) {
 
 				assert.Contains(t, result, "package")
 				assert.Contains(t, result["package"], "github.com/heynemann/go-cov-parser/gocovparser")
-				assert.EqualValues(t, result["package"]["github.com/heynemann/go-cov-parser/gocovparser"], expectedCov)
+				assert.EqualValues(t, expectedCov, result["package"]["github.com/heynemann/go-cov-parser/gocovparser"])
```
Comment (abdullah-arif-swe) [inline gocovparser/core_test.go:44]: This was just backward, creating a wonky test message when it failed

File: gocovparser/core_test.go [hunk 6]
```diff
@@ -122,6 +122,160 @@ func TestCanGroupCoverageData(t *testing.T) {
...
+				assert.Contains(t, result["file"], "github.cbhq.net/some_repo/internal/cmd/server/server.go")
+				assert.EqualValues(t, 0.22900763358778625, result["file"]["github.cbhq.net/some_repo/internal/cmd/server/server.go"])
+				assert.Contains(t, result["file"], "github.cbhq.net/some_repo/internal/config/config.go")
+				assert.EqualValues(t, 0.8490566037735849, result["file"]["github.cbhq.net/some_repo/internal/config/config.go"])
+
+				assert.Contains(t, result, "total")
+				assert.Contains(t, result["total"], "total")
+				assert.EqualValues(
+					t,
+					0.7949806237313157,
```
Comment (Aarif123456) [inline gocovparser/core_test.go:274]: Let'

File: gocovparser/core_test.go [hunk 7]
```diff
@@ -131,6 +285,11 @@ func TestCanGroupCoverageData(t *testing.T) {
 			if !assert.NoError(t, err) {
 				return
 			}
+			defer func() {
+				if t.Failed() {
+					t.Logf("items: %#v\n", items)
```
Comment (abdullah-arif-swe) [inline gocovparser/core_test.go:290]: I found this useful when debugging - lmk if you prefer I remove it

File: gocovparser/core_test.go [hunk 8]
```diff
@@ -172,7 +331,8 @@ func TestCanParseCoverageData(t *testing.T) {
 			expectedData: func(t *testing.T, result []gocovparser.Coverage) {
 				t.Helper()
 
-				assert.Len(t, result, 43)
+				assert.Len(t, result, 1)
+				assert.Len(t, result[0].Blocks, 43)
```
Comment (abdullah-arif-swe) [inline gocovparser/core_test.go:335]: Same idea, just the format is different

File: gocovparser/parsers.go [hunk 9]
```diff
@@ -7,16 +7,16 @@ import (
 // FileParseGroup returns each line as it's own key.
 var FileParseGroup = ParseGroup{
 	Name: "file",
-	KeyFunc: func(line string) string {
-		return strings.Split(line, ":")[0]
+	KeyFunc: func(filename string) string {
+		return filename
```
Comment (abdullah-arif-swe) [inline gocovparser/parsers.go:11]: Just made it work with filename - since that's what the go package provides

File: gocovparser/core.go (hunk 3 above)
Comment (heynemann) [reply gocovparser/core.go:30]: Does this work with go version 1.18+?

**Review by heynemann — APPROVED** [review]
LGTM!!! Thanks for contributing! Left a question but we can merge it assuming we support go 1.19+ (just wondering about 1.18)

**Pull request merged**